[k8 1.24] Use TokenRequest API to get SA token

joe-kimmel-vmw opened the pull request:
#### What this PR does / why we need it:

As part of kubernetes 1.24, a `Secret` is no longer created alongside a serviceaccount. This means that we no longer have access to the `token` generated by the Secret. This PR updates our codepath to use the `TokenRequest` API instead. This has been available (and turned on by default) since kubernetes 1.12 so we should not have any incompatibility from this change. There is no user facing change.

#### Which issue(s) this PR fixes:
<!--
If no issue exists for this change, please create an issue and link it here.
-->
Fixes # https://github.com/vmware-tanzu/carvel-kapp-controller/issues/687

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. 

-->
```release-note
NONE
```

#### Additional Notes for your reviewer:

##### Review Checklist:

- [x] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [x] Relevant tests are added or updated
- [x] Relevant docs in this repo added or updated
- [x] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [x] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

neil-hickey committed:
WIP: Use TokenRequestAPI instead of Secret

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Remove audience from tokenrequest

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
WIP: Use cert from disk instead of configmap

Signed-off-by: Neil Hickey <nhickey@vmware.com>

joe-kimmel-vmw committed:
rbac and dev-deploy fix

joe-kimmel-vmw committed:
kind github action so that we will never again be unkind

joe-kimmel-vmw committed:
who knows how to github action anyhoway

joe-kimmel-vmw committed:
action needs a tmp dir for the e2e tee

joe-kimmel-vmw committed:
kind-action now with secretgen-controller

joe-kimmel-vmw committed:
fix the stupid name of the kind action

neil-hickey committed:
Manage token requests using tokenManager

- This will check the local cache and not hit the API server on every
  run if the token has not expired yet

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Error unable to get token

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Set token expiration to 5 minutes

- remove unused code

Signed-off-by: Neil Hickey <nhickey@vmware.com>

cppforlife commented on pkg/deploy/service_accounts.go:
```diff
@@ -49,43 +56,26 @@ func (s *ServiceAccounts) fetchServiceAccount(nsName string, saName string) (str
 	}
 
-	sa, err := s.coreClient.CoreV1().ServiceAccounts(nsName).Get(context.Background(), saName, metav1.GetOptions{})
+	tokenMgr := token.NewManager(s.coreClient, s.log)
```
since NewManager is made every time we call Find no caching happens afaik.

neil-hickey replied:
yup, I am wondering where it makes sense to have this, perhaps all the way back up the call stack.

Both AppCR and PKGR called `deploy.NewFactory()` so they would both get a different cache, I think having a shared cache is fine.

cppforlife commented on pkg/deploy/service_accounts.go:
```diff
@@ -49,43 +56,26 @@ func (s *ServiceAccounts) fetchServiceAccount(nsName string, saName string) (str
+	})
 	if err != nil {
-		return "", fmt.Errorf("Getting service account: %s", err)
+		return "", fmt.Errorf("Failed to get service account token: %s", err)
```
nit: dont commonly include "Failed" in the error itself (it's an error, so it did indeed fail)

cppforlife commented on pkg/deploy/service_accounts.go:
```diff
@@ -49,43 +56,26 @@ func (s *ServiceAccounts) fetchServiceAccount(nsName string, saName string) (str
+
+	t, err := tokenMgr.GetServiceAccountToken(nsName, saName, &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			ExpirationSeconds: getInt64Point(5000), // 5 minutes
```
a deploy might taken 20min to run as an example, i think our tokens needs to be longer e.g. 1 hour.

cppforlife replied:
nit: instead of having a method getInt64Point you can just set variable to 5000 and take a pointer of it.

cppforlife commented on pkg/deploy/service_accounts.go:
```diff
@@ -49,43 +56,26 @@ func (s *ServiceAccounts) fetchServiceAccount(nsName string, saName string) (str
-		}
-
-		return s.buildKubeconfig(secret)
+	cert, err := os.ReadFile(caCertPath)
```
i dont believe cert is changing so we can in theory cache it after reading it once.

neil-hickey replied:
hmm, I suppose if one ever had to rotate the kube-api cert that all deployments would get restarted anyways? Thus we would re-read from disk in that case.

cppforlife commented on pkg/token/token_manager.go:
```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 VMware, Inc.
```
can we rename this package to `satoken`, token just seems way too generic.

neil-hickey committed:
Create TokenManager during NewServiceAccounts

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Check for caCert on first call only

Signed-off-by: Neil Hickey <nhickey@vmware.com>

joe-kimmel-vmw commented on pkg/satoken/token_manager.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// requiresRefresh returns true if the token is older than 80% of its total
+// ttl, or if the token is older than 24 hours.
```
minor but should update the '24' here to be whatever our token lifetime is.

neil-hickey committed:
Change TTL to 2 hours and refresh to 1 hour

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Garbage collect once a day

- Remove 'expiration' as a cache key - not needed

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Make expiration a constant

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Add tests for TokenManager

- Also pass logger down to TokenManager

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Pass logger to all relevant tests and  code

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Add comments to appease linter

- rename kind action job to reflect it running tests

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey commented on pkg/satoken/token_manager.go:
```diff
@@ -0,0 +1,138 @@
+		clock: clock.RealClock{},
+		log:   log,
+	}
+	go wait.Forever(m.cleanup, gcPeriod)
```
So, I left this in here - but increased the periodic time to a day for garbage collection

I suspect the only time this will be needed is when a service account is deleted / renamed by the user and we still maybe have it in our cache. There is the potential for this to grow without bounds, so I think having this periodic cleanup makes sense in that case.

@cppforlife

cppforlife commented on pkg/deploy/service_accounts.go:
```diff
@@ -49,43 +61,27 @@ func (s *ServiceAccounts) fetchServiceAccount(nsName string, saName string) (str
 
-	for _, secretRef := range sa.Secrets {
-		secret, err := s.coreClient.CoreV1().Secrets(nsName).Get(context.Background(), secretRef.Name, metav1.GetOptions{})
+	if len(s.caCert) == 0 {
```
needs to be behind mutex since parallel calls to Find may/will happen. alternatively can move it into constructor and record err (to be returned from Find) to avoid rereading cert.

cppforlife commented on pkg/deploy/service_accounts.go:
```diff
@@ -4,21 +4,33 @@
 	"k8s.io/client-go/kubernetes"
 )
 
+const (
+	caCertPath      = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+	tokenExpiration = time.Hour * 2
+)
```
can move these consts inside the function. (since it's only needed by that one func)

neil-hickey replied:
`tokenExpiration` feels like a magic number, so I like having it as a constant. The cert location is used once, so I'm fine with that being inlined.

cppforlife replied:
im not suggesting it to be inlined, just be const inside the function e.g.

```
func (Foo) something() {
   const (
     x = 123
   )
   ...
} 
```

cppforlife commented on pkg/satoken/token_manager_test.go:
```diff
@@ -0,0 +1,229 @@
+// This file is a modified version of
+// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/token/token_manager_test.go
+
+package satoken
```
not sure if we can move this into satoken_test

neil-hickey replied:
Lemme give it a shot

neil-hickey replied:
We mock out `getToken` which is a private function of `Manager`. Mocking out the call to k8s is more complicated if we go the `_test` package route.

We'd also have to make the `clock` public which should not be public.

neil-hickey committed:
Put cert read behind a mutex

- Find() can be called in parallel / mutliple times
- Refactor constants to the function where they are used

Signed-off-by: Neil Hickey <nhickey@vmware.com>
